doc/cassbeam.1
.\" Copyright (C) August 18, 2003  Walter Brisken
.\"
.\" %%%LICENSE_START(GPLv2+_DOC_FULL)
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, see
.\" <http://www.gnu.org/licenses/>.
.\" %%%LICENSE_END
.TH "CASSBEAM" "1" "18 Aug 2003" "1.0" "User Commands"
.SH "NAME" 
cassbeam \- Cassbeam is a Cassegrain antenna ray tracer 
.SH "SYNOPSIS" 
.PP 
\fBcassbeam\fR \fIinput_file\fR [\fIkey\fR=\fIvalue\fR ...]
.SH "DESCRIPTION" 
.PP 
\fBcassbeam\fR is a program for Cassegrain antenna modelling.
It computes several properties of the antenna
including gain, zenith system temperature, and the beam, in full
polarization.  All calculations are done in the transmit sense and use
reciprocity to relate to the equivalent receiving system. 
.PP
Cassbeam is a non-interactive command line program that takes all of its input
from the command line.  Note that this does not preclude someone at a later
date from making a graphical or web front end.  There is one required
argument when running cassbeam - the input filename.
Additional arguments can supplement the
parameters of the input file.  These arguments are passed in the 
same \fBkey=value\fR as required in the input file except whitespace
is not allowed around the equal sign.  If a parameter appears both
in the input file and the command line, then the value on the command line
supercedes the value on the input file.
.SH "INPUT FILE"
.PP
The main input file consists of a series of lines of the form 
\fIkey\fR=\fIvalue\fR.  Only one such entry is allowed per line.  The equal 
sign is optional.  The input
files allow comments to be placed within the file.  
All comments begin with \fI%\fR.  This character and any that follow it on
a given line are ignored by cassbeam.  Depending on \fIkey\fR, the \fIvalue\fR
may be one of five types: string, integer, double, vector, none.
A string is a sequence of non-whitespace characters \fInot\fR surrounded 
by quotes of any kind.  A double value is a number that can have a fractional
part.  A vector is a comma-separated list of doubles.  The `none' type expects
no \fIvalue\fR.  Below is a list of
the allowed \fIkeys\fR and the type of \fIvalue\fR expected.  If the range of
legal values is restricted, the legal range will be contained within brackets.
Note that legal values do not imply a physical system that will generate
meaningful results!  For the vector type, if a certain number of values are
needed, they will be indicated in parentheses.  A required parameter will 
be indicated with a `*'.  It is important to realize that
the secondary optical surface (i.e., the subreflector) is defined based on the
input geometry.  Thus changing the feed placement will change the geometry
of the subreflector!  To change parameters of the telescope without affecting
the shape of the subreflector, set the pathology parameters.  Note that the
order of the parameters does not matter.

.SS "Antenna geometry parameters"
.TP
\fBfeed\_x\fR
The x value of the phase center of the feed.  If no value is provided, 0
is assumed. [double]
.TP
\fBfeed\_y\fR
The y value of the phase center of the feed.  If no value is provided, 0
is assumed. [double]
.TP
\fBfeed\_z\fR
The z value of the phase center of the feed.  If no value is provided, 0
is assumed. [double]
.TP
\fBgeom\fR
This string points to a disk file containing the primary optical surface
geometry.  This file is a three column ascii text file, each containing
space separated values for r, z, and dz/dr for the antenna.  There is no
limit (other than your computer's memory) to the number of lines in this file.
It is assumed (but not checked!) that the values of r start at 0 and are
equally spaced.  The radius, R, of the primary is given by the value of r in
the last row.  Columns 1 and 2 are in meters, and column 3 is dimensionless.
[string]
.TP
\fBhole\_radius\fR
The radius (in meters) of an unpanelled area at the center of the primary.  If
omitted, no hole will be made. [double, > 0]
.TP
\fBlegapex\fR
The z value where the legs (struts) intersect each other.  
Note that the legs might terminate before reaching this point.  The default
value is 1.2*\fBsub_h\fR. [double, > 0]
.TP
\fBlegfoot\fR
The r value where the legs (struts) intersect the primary surface.  The
default value is half the antenna radius. [double, > 0] 
.TP
\fBlegwidth\fR
The effective width of the legs, used to compute blockage.  Note that currently
a positive value indicates four equally spaced legs with one leg along the
x axis.  If the value is negative, its absolute value is used in
the blockage calculations, but the legs are rotated 45°.  If this
parameter is not set, or if it is set to 0, then no legs will be generated. [double]
.TP
\fBname\fR
An optional name given to the antenna.  If the name is ``VLBA'', then
the true strut geometry for the VLBA antennas is used rather than 
equispaced struts. [string]
.TP
\fBroughness\fR
The RSS surface roughness in meters.  This number represents the combined
surface error for the primary and secondary.  If no roughness is provided,
the default value of 0 is used. [double, >= 0]
.TP
\fBsub_h\fR
This value is the z value of the intersection of the subreflector with
the z axis. [double, > 0]
.PP
.SS "Feed pattern parameters"
.PP
Note that either both \fBfeedtaper\fR and \fBfeedangle\fR or \fBfeedpattern\fR
must be provided.
.TP
\fBfeedangle\fR
Sets the reference angle for the feed taper. [double, > 0]
.TP
\fBfeedpattern\fR
The name of the file containing the pattern of the feed.  This file contains
two space-separated columns of numbers: the angle in degrees and the taper
in dB.  The first angle must equal 0, and the angles must be uniformly
spaced. [string]
.TP
\fBfeedpatternscale\fR
The factor by which to scale the pattern defined in \fBfeedpattern\fR.
[double, > 0]
.TP
\fBfeedtaper\fR
This parameter sets the taper (in dB) of the feed at an angle \fBfeedangle\fR
from the feed axis to 10^-\fBfeedtaper\fR/10. [double, > 0]
.PP
.SS "Pathology parameters"
.PP
None of the following operations change the shape of the subreflector - its
geometry is calculated before their application.  Note that displacements
of either the feed or the subreflector result in a rotation of the feed
that corrects for the mispointing caused by the translations.  Rotations
of the feed act in addition to this correction.  Composited rotations (i.e.,
setting \fBrsub_x\fR and \fBrsub_y\fR are both provided), the operations
on the object being rotated proceed in reverse alphabetical order (z 
rotation before y rotation; y rotation before x rotation) regardless
of the order that the parameters are received.
.TP
\fBdfeed\_x\fR
Displacement of the feed along the x axis. [double]
.TP
\fBdfeed\_y\fR
Displacement of the feed along the y axis. [double]
.TP
\fBdfeed\_z\fR
Displacement of the feed along the z axis. [double]
.TP
\fBdsub\_x\fR
Displacement of the subreflector along the x axis. [double]
.TP
\fBdsub\_y\fR
Displacement of the subreflector along the y axis. [double]
.TP
\fBdsub\_z\fR
Displacement of the subreflector along the z axis. [double]
.TP
\fBfocus\fR
Displacement of the feed along the feed axis.  A positive value moves the
feed closer to the subreflector. [double]
.TP
\fBrfeed_x\fR
Rotation of the feed in degrees about the x axis.  A positive 
value will rotate from the z axis through the y axis. [double]
.TP
\fBrfeed_y\fR
Rotation of the feed in degrees about the y axis.  A positive 
value will rotate from the x axis through the z axis. [double]
.TP
\fBrfeed_z\fR
Rotation of the feed in degrees about the z axis.  A positive 
value will rotate from the y axis through the x axis. [double]
.TP
\fBrsub_x\fR
Rotation of the subreflector in degrees about the x axis.  A positive 
value will rotate from the z axis through the y axis. [double]
.TP
\fBrsub_y\fR
Rotation of the subreflector in degrees about the y axis.  A positive 
value will rotate from the x axis through the z axis. [double]
.TP
\fBrsub_z\fR
Rotation of the subreflector in degrees about the z axis.  A positive 
value will rotate from the y axis through the x axis. [double]
.TP
\fBsubrotpoint\fR
Defines the point about which the rotation of the subreflector is performed.
The contents of the vector depend on the number of elements are provided:
either only the z value, or the x and y values, or the x, y, and z
values. [vector (1 or 2 or 3)]
.PP
.SS "Operating condition parameters"
.TP
\fBcompute\fR
A string to tell what output to produce.  The string can be `all', `none', 
or a string containing flag characters.  The default value is `all', meaning
produce all possible output.  `none' will produce only messages on the screen
and no output files.  The characters of the general string mean the following:
.IP
\fBa\fR Save the aperture images;
.IP
\fBj\fR Save the Jones matrices in a table;
.IP
\fBp\fR Save the parameters;
.IP
\fBs\fR Save the polarized beams.
.IP
Note that the string is case insensitive. [string]
.TP
\fBdiffeff\fR
A user supplied diffraction efficiency.  If none is provided, an internal
algorithm that is not very good is used.  This needs to be upgraded! [double]
.TP
\fBfreq\fR
The frequency in GHz at which the calculation will be run. [double, > 0]
.TP
\fBgridsize\fR
Specifies a fixed grid size.  If odd, the next even number will be used.
This option overrides any setting of \fBoversamp\fR and is the preferred 
method of setting the grid size.  Setting it to a value less than 32 will
result in a grid size of 32. [integer, >= 32]
.TP
\fBleggroundscatter\fR
The fraction of power that scatters off the struts toward the ground.  The
default value is 0.2. [double, >= 0, <= 1]
.TP
\fBmisceff\fR
A factor of the efficiency calculation that contains ``everything else''.  
The user is responsible for choosing a realistic value for this.  A default
of 1 (i.e., 100%) is assumed if this parameter is not provided.
[double, >= 0, <= 1]
.TP
\fBout\fR
The prefix for all output files.  The default is \fIcassbeam\fR.  A dot
will always separate the prefix from any trailing characters. [string]
.TP
\fBoversamp\fR
One way of specifying the grid size.  This option will make the grid on the
primary fine enough to accommodate 4*\fBoversamp\fR*R/lambda points.  The
default is 1.  Note that vastly ``undersampling'' is fine as the field is
never calculated anywhere between the feed and the aperture plane.  Normally
blockage calculations and constancy of the illumination will dictate the
required sampling.  See \fBgridsize\fR for an alternate way of specifying
the grid.  This parameter is ignored if \fBgridsize\fR is set. [double, > 0]
.TP
\fBpixelsperbeam\fR
This is the approximate number of pixels that the core of the beam will
occupy in the output images. [int, > 0]
.TP
\fBTground\fR
The temperature in Kelvin of the ground.  The default value is 290. [double, > 0]
.TP
\fBTrec\fR
The equivalent temperature of the receiver.  This adds into the system
temperature.  The default value is 50. [double, > 0]
.TP
\fBTsky\fR
The temperature in Kelvin of the sky.  The default value is 3 for frequencies
over 1 GHz, and 3 * 10^-2.5 nu for frequencies below 1 GHz. [double, > 0]
.PP
.SH "OUTPUT FILES"
.PP
Up to 12 output files are generated depending
on which \fBcompute\fR options were selected at run time.  These files are
listed below.  The letter in brackets in the section headings indicate which
option is used to enable this file to be written.  All output files begin
with the value of the input parameter \fBout\fR.  Currently all output images
are in PGM format, which is a very simple greyscale 
image format supported by most unix-based image viewers.  
.SS "Aperture images [a]"
.PP
Three images are generated that allow the aperture field to be examined
qualitatively.
If quantitative numbers are needed, the source code should be modified to
export the illumination parameters.
.TP
\fIout\fB.illumamp.pgm\fR
Raster image showing the amplitude
of the illumination pattern of the primary.  No blockage is done at this
point.  The scale is linear in flux.
.TP
\fIout\fB.illumphase.pgm\fR
Raster image showing the
net phase (pathlength multiplied by wave vector) at each point on the
primary.  A phase gradient is removed.  Portions of the image that correspond
to zero flux have an arbitrary phase.
.TP
\fIout\fB.illumblock.pgm\fR
Raster image showing the
blocked portion of the aperture.  White means that this part of the dish is 
experiences either plane wave blockage from the
sky or spherical wave blockage from the feed, and thus does not contribute
to the gain of the antenna.
.PP
.SS "Jones matrix file [j]"
.PP
The Jones matrix file, \fIout\fB.jones.dat\fR contains the Jones matrix
(see Hamaker et al. 1996 for 
details) corresponding to the effect of the antenna on the incoming
radiation as a function of position on the sky.  The file is organized as
an eight column ascii.  The first row corresponds
to the point on the image with smallest l and m.  The rastering
then proceeds first with increasing l, and then with increasing m.  
There are a total of n^2 rows, where n is the smallest odd number
greater than or equal to the \fBgridsize\fR used.  The matrices are
rastered on a sine-projected coordinate system tangent to the sky at the
beam center, which corresponds to row number (n^2+1)/2.  At the beam center
the pixel scale is given by the output parameter \fBbeampixelscale\fR, which
is stored in the output file \fIout\fB.params\fR described below.
.PP
.SS "Parameter file [p]"
.PP
The parameter file, \fIout\fB.params\fR is an output file in the same format
as the input file, containing all of the input parameters that were 
specified (even if on the command line) as well as many output 
values.  They are:
.TP
\fBAeff\fR
The effective area of the antenna [m^2]. [double]
.TP
\fBAeff\_Tsys\fR
The effective area of the antenna divided by the system temperature
[m^2/K]. [double]
.TP
\fBampeff\fR
The amplitude efficiency. [double]
.TP
\fBbeampixelscale\fR
The scale of the generated beam images [deg/pixel]. [double]
.TP
\fBblockeff\fR
The blockage efficiency. [double]
.TP
\fBdiffeff\fR
The diffraction efficiency. [double]
.TP
\fBfwhm\_l\fR
The full width at half max of the beam in the l direction. [double]
.TP
\fBfwhm\_m\fR
The full width at half max of the beam in the m direction. [double]
.TP
\fBgain\fR
The gain G of the antenna. [double]
.TP
\fBillumeff\fR
The illumination efficiency. [double]
.TP
\fBpeaksidelobe\fR
The directivity of the greatest sidelobe relative to the peak directivity
of the beam. [double]
.TP
\fBphaseeff\fR
The phase efficiency. [double]
.TP
\fBpoint\_l\fR
The l component of the pointing offset from the z axis 
measured in the image plane. [double]
.TP
\fBpoint\_m\fR
The m component of the pointing offset from the z axis 
measured in the image plane. [double]
.TP
\fBprispilleff\fR
The primary spillover efficiency. [double]
.TP
\fBprogram\fR
The name of the program run, which is \fIcassbeam\fR. [string]
.TP
\fBmisceff\fR
The miscellaneous efficiency. [double]
.TP
\fBspilleff\fR
The spillover efficiency. [double]
.TP
\fBsubspilleff\fR
The subreflector spillover efficiency. [double]
.TP
\fBsurfeff\fR
The surface efficiency. [double]
.TP
\fBtotaleff\fR
The total efficiency calculated for the antenna. [double]
.TP
\fBTsys\fR
The system temperature. [double]
.TP
\fBversion\fR
The software version number. [string]
.PP
.SS "Polarized beam images [s]"
.PP
With the \fBs\fR option, cassbeam will produce 7 images of the beam showing
in the four Stokes parameters the response to an unpolarized source as
a function of the position of the source on the sky.  This information
is derived from the Jones matrices which are saved in
\fIout\fB.jones.dat\fR. These images are meant for qualitative inspection.
The Jones matrices contain the formal output.
.TP
\fIout\fB.I.pgm\fR
Stokes I - total intensity; 
.TP
\fIout\fB.Q.pgm\fR
Stokes Q - excess linear polarization e_1 over e_2;
.TP
\fIout\fB.U.pgm\fR
Stokes U - excess linear polarization in e'_1 over e'_2
.TP
\fIout\fB.V.pgm\fR
Stokes V - excess right circular polarzation
over left circular polarization;
.TP
\fIout\fB.QI.pgm\fR
The ratio of the Stokes Q image to the 
Stokes I image;
.TP
\fIout\fB.UI.pgm\fR
The ratio of the Sytokes U image to the 
Stokes I image; 
.TP
\fIout\fB.VI.pgm\fR
The ratio of the Stokes V image to the 
Stokes I image; 
.PP
.SH "AUTHOR"
.PP
\fBCassbeam\fR is written by  Walter Brisken, National Radio Astronomy
Observatory. This manpage is extracted from his cassbeam manual.
.SH "SEE ALSO"
.PP
See the complete manual in /usr/share/doc/cassbeam/ for more information. 


CHANGES
1.1
===

Changes since 1.0

 * make compile on modern systems
 * don't generate doc by default
 * support for DESDIR in Makefile
 * Add manual page



cassbeam.c
/* cassbeam - a Cassegrain antenna simulator
    Copyright (C) August 18, 2003  Walter Brisken

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, see
    <http://www.gnu.org/licenses/>.
 */
#include <stdio.h>
#include <math.h>
#include <glib.h>
#include <string.h>
#include "vector.h"
#include "image.h"
#include "image-vector.h"
#include "vector-fftw.h"
#include "vecarray.h"
#include "polygon.h"
#include "randdist.h"
#include "antenna.h"
#include "illum.h"

const char program[] = "cassbeam";
const char version[] = "1.0";
const char versiondate[] = "08/18/2003";

/*  TODO
 *
 *
 * POST 1.0
 *   Add opacity for gain and Tsys calculations
 *   Compute phase gradient in primary beam (DFT beam cuts)
 *   Oversample blockage calculations?
 *   More accurate diffraction loss?
 *   Feed radius/azimuth instead of x,y
 */

/* given an array of alternating re and im components, mult by exp(i phase) */
void phaseshift(Vector e, double phase)
{
	double c, s;
	Vector E;
	int i, N;

	N = VectorSize(e);
	g_assert(N % 2 == 0);

	c = cos(phase);
	s = sin(phase);

	E = dupVector(e);

	for(i = 0; i < N; i+=2)
	{
		e[i  ] = c*E[i] - s*E[i+1];
		e[i+1] = s*E[i] + c*E[i+1];
	}

	deleteVector(E);
}

/* Returns 4 matrices : Re[Ex], Im[Ex], Re[Ey], Im[Ey] */
Illum *cassillum(const Antenna *a, const Pathology *p, 
	const struct KeyValue *kv)
{
	int i, j, ii, jj, N, n, iter, niter=7;
	double dx, dA, R2, H2, eps;
	Ray *ray, *rayx, *rayy, *bray;
	double x, y, r2, x1, y1, bx, by;
	double dx1, dy1, dx2, dy2;
	double dO, dP;
	double amp, L, L0;
	Vector E, Er, El, Pr, Pl;
	Illum *I;
	double Pforward, Pbackward, Pspill, Pground, Pleg=0.0, Phole=0.0;
	double Psub=0.0, Ppri=0.0, Ptot;
	double phase, cp, sp;
	double subperimeter;
	const int nrim=120;
	VecArray subrim;  /* columns are : X, Y, Z, theta, rho, dtheta */
	VecArray B;

	/* compute total power coming out of the feed */
	Pforward = feedpower(a);
	Pbackward = feedbackpower(a);
	Ptot = Pforward + Pbackward;

	/* compute rim of subreflector and the total power hitting it */
	subrim = calcsubreflectorrim(a, p, nrim);
	Psub = subpower(subrim, a);
	subperimeter = polygonperimeter(subrim[0], subrim[1]);
	
	/* compute image size */
	if(a->gridsize > 0) n = a->gridsize/2;
	else n = 2.0*a->oversamp*a->radius/a->lambda + 0.999;
	if(n%2 == 1) n++;
	N = 2*n;
	
	dx = a->radius/n;
	dA = dx*dx;
	eps = dx/9.237;
	R2 = a->radius*a->radius;
	H2 = a->hole_radius*a->hole_radius;

	/* compute central ray pathlength */
	ray = trace(a, 0.0, 0.00001, p);
	L0 = Raylen(ray);
	deleteRay(ray);

	/* compute polarization vectors */
	Pr = newVector(4);
	Pl = newVector(4);
	Pr[0] = 1.0/M_SQRT2; Pr[1] = 0.0; Pr[2] = 0.0; Pr[3] = 1.0/M_SQRT2;
	Pl[0] = 1.0/M_SQRT2; Pl[1] = 0.0; Pl[2] = 0.0; Pl[3] = -1.0/M_SQRT2;
	Er = Efield(a, p, Pr);
	El = Efield(a, p, Pl);
	deleteVector(Pr);
	deleteVector(Pl);

	B = getfeedbasis(a, p);

	/* Initialize the illumination structure */
	I = newIllum(N);
	setIllumfromKeyValue(I, kv);
	I->dA = dA;
	I->a0 = a->radius*a->radius*M_PI;
	I->lambda = a->lambda;
	I->Tsky = a->Tsky;
	I->Tground = a->Tground;
	I->Trec = a->Trec;
	I->Xangle = atan2(B[0][1], B[0][0]);
	I->gridsize = a->gridsize;
	
	phaseshift(Er, -I->Xangle);
	phaseshift(El, I->Xangle);
	I->Xangle = 0.0;
	
	/* compute field on primary */
	for(j = -n; j < n; j++)
	{
		y = (j + 0.5)*dx;
		jj = j + n;
		for(i = -n; i < n; i++) 
		{
			x = (i + 0.5)*dx;
			ii = i + n;
			ray = rayx = rayy = bray = 0;
			
			r2 = x*x + y*y;
			if(r2 > R2) continue;

			x1 = x;
			y1 = y;

			/* Iterate to find starting point */
			for(iter = 0; iter < niter; iter++)
			{
				ray = trace(a, x1, y1, p);
				if(!ray) goto nextpoint;
				x1 += (x - ray->aper[0]);
				y1 += (y - ray->aper[1]);
				deleteRay(ray);
				ray = 0;
				if(x1*x1 + y1*y1 > R2) goto nextpoint;
			}
	
			if(y < 0) rayy = trace(a, x1, y1+eps, p);
			else rayy = trace(a, x1, y1-eps, p);
			
			if(x < 0) rayx = trace(a, x1+eps, y1, p);
			else rayx = trace(a, x1-eps, y1, p);
			
			ray = trace(a, x1, y1, p);
			
			if(ray == 0 || rayx == 0 || rayy == 0)
				goto nextpoint;

			dx1 = rayx->aper[0]-ray->aper[0];
			dy1 = rayx->aper[1]-ray->aper[1];
			dx2 = rayy->aper[0]-ray->aper[0];
			dy2 = rayy->aper[1]-ray->aper[1];

			/* To make blockage statistically correct... */
			bx = x + 0.5*rand_pm_one()*(dx1/eps)*dx;
			by = y + 0.5*rand_pm_one()*(dy2/eps)*dx;
			bray = trace(a, bx, by, p);

			dA = 0.5*fabs(dx1*dy2 - dx2*dy1);

			dO = (dOmega(a, rayx, rayy, ray, p)/dA)*dx*dx;
			dP = dO*feedgain(a, ray, p);

			/* Check for blockage */
			if(bx*bx+by*by < H2) /* hole in the middle */
			{
				I->B[jj][ii] = 1.0;
				Phole += dP;
			}
			else if(legplanewaveblock2(a, bray) ||   /* legs */
				legsphericalwaveblock(a, bray))
			{
				I->B[jj][ii] = 1.0;
				Pleg += dP;
			}
			
#if 0
			if(polygonside(subrim[0], subrim[1], 
				ray->sub[0], ray->sub[1]) < 0)
			{
				Po += dP;
				printf(".");
				goto nextpoint;
			}
#endif

			Ppri += dP;

			/* Fill in Scalar field matrices */
			amp = sqrt(dP);
			L = Raylen(ray);
			phase = 2.0*M_PI*(L-L0)/a->lambda;
			sp = sin(phase);
			cp = cos(phase);
			I->A[jj][ii] = amp;
			I->P[jj][ii] = phase;
			
			/* Fill in Vector field matrices */
			E = tracepol(Er, ray);
			I->E[RXR][jj][ii] = amp*(E[0]*cp - E[1]*sp);
			I->E[IXR][jj][ii] = amp*(E[0]*sp + E[1]*cp);
			I->E[RYR][jj][ii] = amp*(E[2]*cp - E[3]*sp);
			I->E[IYR][jj][ii] = amp*(E[2]*sp + E[3]*cp);
			I->E[RZR][jj][ii] = amp*(E[4]*cp - E[5]*sp);
			I->E[IZR][jj][ii] = amp*(E[4]*sp + E[5]*cp);
			deleteVector(E);
			E = tracepol(El, ray);
			I->E[RXL][jj][ii] = amp*(E[0]*cp - E[1]*sp);
			I->E[IXL][jj][ii] = amp*(E[0]*sp + E[1]*cp);
			I->E[RYL][jj][ii] = amp*(E[2]*cp - E[3]*sp);
			I->E[IYL][jj][ii] = amp*(E[2]*sp + E[3]*cp);
			I->E[RZL][jj][ii] = amp*(E[4]*cp - E[5]*sp);
			I->E[IZL][jj][ii] = amp*(E[4]*sp + E[5]*cp);
			deleteVector(E);

		nextpoint:
			if(ray)  deleteRay(ray);
			if(rayx) deleteRay(rayx);
			if(rayy) deleteRay(rayy);
			if(bray) deleteRay(bray);
		}
	}

	deleteVecArrayandVectors(B);
	deleteVecArrayandVectors(subrim);
	deleteVector(Er);
	deleteVector(El);

	/* remove phase gradient, populate pointing fields */
	dephaseIllum(I);
	/* again, with incremental inprovements */
	dephaseIllum(I);

	Pspill = Ptot - Ppri;

	Pground = Pspill - (Ptot-Psub) + a->leggroundscatter*Pleg;

	I->groundfrac = Pground/Ptot;
	I->spilleff = 1.0-Pspill/Ptot;
	I->subspilleff = Psub/Ptot;
	I->prispilleff = I->spilleff/I->subspilleff;
	x = 4.0*M_PI*a->roughness/a->lambda;
	I->surfeff = exp(-x*x);
	/* Emperical -- FIXME */
	if(I->diffeff == 0.0)
		I->diffeff = 1.0-5.5*a->lambda/subperimeter;

	calcIllumparams(I);

	return I;
}

void savecassdata(const Antenna *a, const Illum *I)
{
	char filename[1000];

	if(I->saveapertureimages)
	{
		sprintf(filename, "%s.illumamp.pgm", I->prefix);
		saveMatrixaspgm(I->A, filename);
		sprintf(filename, "%s.illumphase.pgm", I->prefix);
		saveMatrixaspgm(I->P, filename);
		sprintf(filename, "%s.illumblock.pgm", I->prefix);
		saveMatrixaspgm(I->B, filename);
	}
}

int makecassbeam(struct KeyValue *kv, const char *paramfile)
{
	Pathology *p;
	Antenna *a;
	Illum *I;
	char filename[1000];
	
	a = newAntennafromKeyValue(kv, paramfile);
	g_assert(a);
	p = newPathologyfromKeyValue(kv);
	g_assert(p);

	alignfeed(a, p);

	printAntenna(a);
	printPathology(p);

	if(p->focus != 0.0) defocusAntenna(p, a);

	I = cassillum(a, p, kv);
	printIllum(I);
	calcIllumpolparams(I);
	
	savecassdata(a, I);
	
	KeyValueupdateIllum(kv, I);

	if(I->saveparams)
	{
		sprintf(filename, "%s.params", I->prefix);
		saveKeyValue(kv, filename);
	}
	
	deleteIllum(I);

	deletePathology(p);
	deleteAntenna(a);

	return 1;
}

int main(int argc, char **argv)
{
	struct KeyValue *kv;

	int i, j;
	char k[100], v[100];

	if(argc < 2)
	{
		fprintf(stderr, "%s version %s\n\n", program, version);
		fprintf(stderr, "Walter Brisken, NRAO, %s\n\n", versiondate);
		fprintf(stderr, "usage : %s <filename> [options]\n", argv[0]);
		return 0;
	}

	kv = loadKeyValue(argv[1]);
	if(!kv) return 0;
	
	if(argc > 2) for(i = 2; i < argc; i++)
	{
		if(argv[i][0] == '=') fprintf(stderr,
			"Warning -- option begins with =\n");
		for(j = 0; argv[i][j] != 0; j++) if(argv[i][j] == '=')
			argv[i][j] = ' ';
		if(sscanf(argv[i], "%s %s", k, v) != 2) continue;
		
		j = KeyValuekeyindex(kv, k);
		if(j < 0) KeyValueaddparm(kv, k, v);
		else
		{
			g_free(kv->value[j]);
			kv->value[j] = g_strdup(v);
		}
	}

	/* force some default values */
	if(KeyValuekeyindex(kv, "out") < 0)
		KeyValueupdateparm(kv, "out", program);
	if(KeyValuekeyindex(kv, "compute") < 0)
		KeyValueupdateparm(kv, "compute", "all");
	KeyValueupdateparm(kv, "program", program);
	KeyValueupdateparm(kv, "version", version);

	makecassbeam(kv, argv[1]);

	deleteKeyValue(kv);

	printf("\n%s has finished\n", program);

	return 0;
}
				
